### PR TITLE
Allow SharedInformerOptions to be passed to dynamic informers

### LINKED ISSUE #1066: Add NewFilteredDynamicInformerWithOptions to client-go dynamic/dynamicinformer
Also add NewFilteredDynamicSharedInformerFactoryWithOptions

Part of #981

### PR DESCRIPTION
**Pull request by fabianvf**
Closes https://github.com/kcp-dev/kcp/issues/1066

### CODE REVIEW

File: staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go [hunk 1]
```diff
@@ -40,13 +40,25 @@ func NewDynamicSharedInformerFactory(client dynamic.Interface, defaultResync tim
...
+		namespace,
+		tweakListOptions,
+		cache.WithResyncPeriod(defaultResync),
+		cache.WithIndexers(cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}),
+	)
+}
+
+// NewFilteredDynamicSharedInformerFactoryWithOptions constructs a new instance of dynamicSharedInformerFactory.
+// Listers obtained via this factory will be subject to the same filters as specified here.
+func NewFilteredDynamicSharedInformerFactoryWithOptions(client dynamic.Interface, namespace string, tweakListOptions TweakListOptionsFunc, opts ...cache.SharedInformerOption) DynamicSharedInformerFactory {
```
Comment (ncdc) [inline staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go:54]: (no action required, can do in a follow up (or not)) do you think `tweakListOptions` should be removed and become a `cache.SharedInformerOption` func?

**Review by ncdc — APPROVED** [review]

**Comment by ncdc** [comment]
Ready to merge if you are @fabianvf ?

**Comment by fabianvf** [comment]
@ncdc if you're good with it so am I, should the commit have the upstream marker in it (since presumably we're going to want this in the upstream PR as well)?

**Comment by ncdc** [comment]
Yes, please adjust the commit message, thanks!

File: staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go (hunk 1 above)
Comment (fabianvf) [reply staging/src/k8s.io/client-go/dynamic/dynamicinformer/informer.go:54]: I don't think so, the `tweakListOptions` are used by the `NewFilteredDynamicInformerWithOptions` when constructing the listwatcher that is passed into `cache.NewSharedIndexInformerWithOptions`

**Pull request merged**